### ISSUE TITLE
Suggestion: Use typeof of declaration to define its type

### ISSUE THREAD

**Issue by DeyLak** [microsoft/TypeScript#32694]
## Search Terms

- infer variable declaration
- typeof declaration
- using typeof on current expression
- apply modifier to declaration 
- generic declarations 

## Suggestion

I would like to use current declaration type for modifying type of variable, I'm creating, e.g.:
```typescript
const foo: Readonly<typeof var> = {
  bar: 'baz',
}
```
Maybe, there is a better syntax, that I couldn't think of.
## Use Cases

Typescript infer rather narrow type from variable declaration by default, which is a good thing. Sometimes we want to widen it, so it can fit into our interfaces. Mostly, it can be done by simply declaring a variable of certain type. But sometimes, it's useful to create some object along the way and somehow extend it's inferred type(some libraries can export very complicated generic types, so explicitly writing them could be a pain)

My specific use case was caused by reselect library, when I tried to reuse some selectors.

```typescript
// typeof selectors is (selector1ReturnType | selector2ReturnType | selector3ReturnType)[]
const selectors = [
  state => someSelector1(state),
  state => someSelector2(state),
  state => someSelector3(state),
]
```
But this array of selectors couldn't be used in `createSelector` function, cause it is expecting a tuple of selectors, to correctly infer type of arguments of selector function. I tried to use `as const` on my declaration, but it leaded to another error, cause createSelector expects a mutable tuple(it's only a type thing, not that it really mutates the selectors, but I'm afraid, this is a really common case in all library typings, not to mark it's arguments as readonly). I also found [a solution](https://gist.github.com/jcalz/381562d282ebaa9b41217d1b31e2c211) for creating tuples with function, but I don't like the fact, that static typings require some(even tiny) runtime overhead, also the solution is not generic.
For now, I ended up with something like this, using `Writable` type from [this](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-4.html) docs:
```typescript
const selectorsConst = [
  state => someSelector1(state),
  state => someSelector2(state),
  state => someSelector3(state),
] as const
const selectors: Writable<typeof selectorsConst> = selectorsConst as any
```
But this solution is not pretty also. I wanted it to look like this:
```typescript
const selectorsConst: Writable<typeof var> = [
  state => someSelector1(state),
  state => someSelector2(state),
  state => someSelector3(state),
] as const
```
Or even something like this(but this is out of scope of this issue):
```typescript
const selectorsConst: Tuple<typeof var> = [
  state => someSelector1(state),
  state => someSelector2(state),
  state => someSelector3(state),
]
```
## Examples

Making a Readonly, Partial, Writable, etc. object:
```typescript
const foo: Partial<typeof var> = {
  bar: {
    baz: 'value',
  },
}
foo.bar.baz = 'new value' // Error: Object is possibly 'undefined'
```
Unifying object values types:
```typescript
interface MyClass {
  bar?: string;
  baz?: string;
}
const foo: {
  [key in keyof typeof var]: MyClass;
} = {
  firstEntry: { bar: 'bar' },
  secondEntry: { baz: 'baz' },
}
```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by krryan** [comment]
I’m not 100% sure I understand, but does `as const` cover what you need? It looks like it should. (Unfortunately for arrays `as const` uses `ReadonlyArray` which isn’t compatible with `Array`, though).

**Comment by RyanCavanaugh** [comment]
Let me attempt to summarize this and also suggest some alternate syntax:
```ts
interface Pair<T> {
  a: T;
  b: T;
}

// Same as if I had written p: Pair<number>
const p: Pair<infer> = {
  a: 10,
  b: 20
};
```

**Comment by goodmind** [comment]
I think it is more along the lines of
```ts
interface Pair<T> {
  a: T;
  b: T;
}

// Same as if I had written p: Pair<{a: number, b: number}>
const p: Pair<infer> = {
  a: 10,
  b: 20
};
```

**Comment by RyanCavanaugh** [comment]
Can you fix your example? There's an error there and I'm not sure in which direction you meant it to be.

**Comment by goodmind** [comment]
@RyanCavanaugh where?

**Comment by goodmind** [comment]
it wouldn't really work, there would be type-error, it is probably meant for idempotent operations

**Comment by DeyLak** [comment]
@RyanCavanaugh that's not what I meant. The type of `infer` should be exactly the type, typescript would assign to your declaration. In your example it would be:
```
{
  a: number;
  b: number;
}
```
So, `Pair` interface would not receive `number` for `T`

Probably, you could change your example to something like:
```
interface Pair<T> {
  a: T;
  b: T;
}

// Same as if I had written p: Pair<{a: number, b: number}>
const p: Pair<infer['a']> = {
  a: 10,
  b: 20
};
```

**Comment by DeyLak** [comment]
@goodmind yes, I think, that the main case for this is idempotent operations on types, like Partial, Readonly, ReturnType etc. Maybe there are more uses.

**Comment by DeyLak** [comment]
I found another use case for this suggestion: 
```typescript
// Use type described here https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html to achieve type parameter inference
type AnyFunction = (...args: any[]) => any
type TargetInterface<T> = {
  field: T, 
  otherField: boolean;
}
type FunctionThatShouldReturnInterface<T extends AnyFunction> = T extends (...args: any[]) => 
TargetInterface<infer R>
  ? () => TargetInterface<R>
  : never

// Use my suggestion here
const getInterface: FunctionThatShouldReturnInterface<infer> = () => ({
  field: someComplexTypeSelector(), // Type, that we don't want to repeat, cause it can be complex
  otherField: true;
})

// Without suggestion, and it can get more complex, if we use more fields
const getInerface2 = (): TargetInterface<ReturnType<typeof someComplexTypeSelector>> => ({
  field: someComplexTypeSelector(), 
  otherField: true;
})

// Same hack, as I described in the main post
const getInerfaceConst = () => ({
  field: someComplexTypeSelector(), 
  otherField: true;
})
export const getInerface3: FunctionThatShouldReturnInterface<typeof getInerfaceConst> = getInerfaceConst
```

#### The real problem, that I tried to solve was like this: 
1. I have a React SPA, that uses hooks, to load data in components, that represent different pages. Obviously, each page has different data format
2. I want to use that hooks on app load(in the main component), so all the pages can load their data in background
3. Data load hooks return data in unified tuple format: `[data: any, isLoading: boolean]`
4. I want make type for this hook, so it can check, if user returns required tuple, but I don't want to force him to write the type of data, that he returns explicitly, cause it can be easily inferred from function ReturnType.

@RyanCavanaugh Maybe you can give me some directions, if you think, this can be implemented without big overhead, so I can make a PR? I'm new to ts codebase, so any help will be appreciated.
